sympy/vector/functions.py

```diff
@@ -196,24 +216,68 @@ def gradient(scalar, coord_sys):
     scalar : SymPy Expr
         The scalar field to compute the gradient of
 
-    coord_sys : CoordSysCartesian
-        The coordinate system to calculate the gradient in
-
     Examples
     ========
 
-    >>> from sympy.vector import CoordSysCartesian, gradient
-    >>> R = CoordSysCartesian('R')
+    >>> from sympy.vector import CoordSystem3D, gradient
+    >>> R = CoordSystem3D('R')
     >>> s1 = R.x*R.y*R.z
-    >>> gradient(s1, R)
+    >>> gradient(s1)
     R.y*R.z*R.i + R.x*R.z*R.j + R.x*R.y*R.k
     >>> s2 = 5*R.x**2*R.z
-    >>> gradient(s2, R)
+    >>> gradient(s2)
     10*R.x*R.z*R.i + 5*R.x**2*R.k
 
     """
+    # Check input
+    if isinstance(scalar, Zero) or scalar == 0:
+        return Vector.zero
+
+    if not issubclass(type(scalar), Expr) or scalar.is_Vector:
+        raise TypeError("Your input should be a scalar field.")
+
+    for atom in scalar.atoms():
+        if isinstance(atom, BaseScalar):
+            system = atom.system
+            break
```

Comment:
> 1. Should we keep CoordSysCartesian() as a subclass of an add superclass CoordSystem()?

No, this would be a violation of the DRY principle, and a useless repetition of code.

> 2. Do we want to restrict the CoordSystem() to Euclidian space?

Yes, for non-Euclidean geometries such as Riemannian and pseudo-Riemannian geometries there's _sympy.diffgeom_, which is specifically enginnered for that.

Typical operators of vector calculus have to be defined in a completely different way in pure differential geometry.

> 3. Have the superclass CoordSystem(R=3, sys=spherical) build the differential class within.

Use classes to correspond to mathematical objects. I would strongly avoid using two levels of classes to represent the same mathematical concept. You should create subclasses if you are restricting to a specific subset of coordinate systems which would require more methods to add to the class. If not so, DO NOT CREATE SUBCLASSES.